keps/1432-admission-check-per-resource-flavor/README.md

```diff
@@ -0,0 +1,94 @@
+# KEP-1432: AdmissionChecks per ResouceFlavor
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This KEP introduces assigning AdmissionChecks to a ResourceFlavor.
+
+## Motivation
+
+Currently, AdmissionChecks are only assigned to a ClusterQueue. This results in running `AdmissionChecks` regardless of
+the ResourceFlavor used by a Workload, even when it is unnecessary. We would like to have the ability to assign AdmissionChecks
+to a specific ResourceFlavor, so the whole mechanism is more flexible and expressive.
+
+### Goals
+- Support running AdmissionChecks only when a particular ResourceFlavor is assigned to the Workload.
+- Running an AdmissionCheck with different settings depending on the flavor
+
+### Non-Goals
+
+## Proposal
+
+Add a new field to FlavorQuota API, so it contains a list of AdmissionChecks that  should run when a Workload
+is admitted for this ResourceFlavor.
+
+### User Stories (Optional)
+
+#### Story 1
+As a user who has reserved machines at my cloud provider I would like to use them first. If they are not obtainable then
+switch to spot machines, using ProvisioningRequest. It only makes sense to run Provisioning AdmissionCheck when the ResourceFlavor is Spot.
+
+### Notes/Constraints/Caveats (Optional)
+User cannot define AdmissionChecks both at ClusterQueue and ResourceFlavor level. This will be validated by
+a ClusterQueue webhook.
+
+### Risks and Mitigations
+
+## Design Details
```

Comment:
I think that this API design would bring AdmissionChecks fragmentation.
So, I'm wondering if we can add configurations under the `ClusterQueueSpec` field like this.
In theory, once we introduce the `AdmissionCheckStrategy`, the existing `AdmissionChecks` field is not needed, but we keep the existing field for backward compatibility.

@alculquicondor @PBundyra WDYT?

```diff
type ClusterQueueSpec struct {
...
    AdmissionChecks []string
+   AdmissionCheckStrategy []AdmissionCheckStrategy
}

+type AdmissionCheckPolicy struct {
+    // AdmissionCheck name.
+    Name string
+    Strategy AdmissionCheckStrategyKind
+    Selector  *metav1.LabelSelector OR []types.NamespacedName
+}

+type AdmissionCheckStrategyKind string
+constants (
+    AdmissionCheckStrategyKindClusterQueue AdmissionCheckStrategyKind = "ClusterQueue" // default
+    AdmissionCheckStrategyKindResourceFlavor AdmissionCheckStrategyKind = "ResourceFlavor"
+)
+
```

Reply:
What is the label selector for? Is it for workload labels?

I'm failing to see what would be the use case at the moment, but I agree that we might need to add more "options" in the future.

OTOH, just a `strategy` is not enough. Maybe it could be something like:

```diff
type ClusterQueueSpec struct {
...
    AdmissionChecks []string
+   AdmissionCheckStrategy []AdmissionCheckStrategy
}

+type AdmissionCheckPolicy struct {
+    // AdmissionCheck name.
+    Name string
+    ForFlavors []string
+}
```

And if `forFlavors` is empty, then the admission check applies to all of the ClusterQueue.

Reply:
The advantage of the above API, compared to the original proposal, is that it's more future proof.

Reply:
> What is the label selector for? Is it for workload labels?

It is for ResourceFlavor, but we can use it for Workloads in the future.
But I don't have a strong point of view against LabelSelector, and I just wanted to express somehow generic filters.

> I'm failing to see what would be the use case at the moment, but I agree that we might need to add more "options" in the future.

Yes, that's right. That is the main motivation for my proposed API.

> OTOH, just a strategy is not enough. Maybe it could be something like:
And if forFlavors is empty, then the admission check applies to all of the ClusterQueue.

I like your proposed API rather than my proposal, but the `ForFlavors` sounds slightly weired.
How about the `CheckForFlavors` or `ActionForFlavors`?

Also, should we limit the number of `forFlavors`? I guess it is possible to set `16`.
WDYT?

Reply:
How about this?
```yaml
admissionCheckStrategies:
- name: foo
  onFlavors: [bar]
```

I think more words is just redundant.

Reply:
it looks great! Thank you!